src/include/kernel/header.h

```diff
@@ -36,4 +36,26 @@ struct sof_abi_hdr {
 	uint32_t data[0];	/**< Component data - opaque to core */
 } __attribute__((packed));
 
+/**
+ * struct sof_ipc4_abi_hdr - Used by any bespoke component data structures or binary blobs.
+ * @magic: The magic number for the header. The value is 'S', 'O', 'F', '4'
+ * @size: The size in bytes of the data, excluding this struct
+ * @abi: SOF ABI version
+ * @blob_type: Type of blob: INIT_INSTANCE, CONFIG_SET or LARGE_CONFIG_SET
+ *             The value is of type enum sof_ipc4_module_type
+ * @param_id: ID indicating which parameter to update with the new data. The validity of
+ *            the param_id with blob_type is dependent on the module implementation.
+ * @reserved: Reserved for future use
+ * @data: Component data - opaque to core
+ */
+struct sof_ipc4_abi_hdr {
```

Comment:
this looks very similar to ipc3 version, can we use the same structure and use magic to determine where to treat things differently

Reply:
Ack on that. It seems "type" can be used differently based on "abi" value, and the first reserved 32bit word can be used as param_id (ignored when abi is IPC3 or older).

Reply:
We removed ```type``` in the struct sof_abi_hdr. So the member offset is different between sof_abi_hdr and sof_ipc4_abi_hdr. In this case, how can we use the same structure to refer the member, for example, ```size```?
sof_abi_hdr->size offset is 8 while sof_ipc4_abi_hdr->size offset is 4.

Reply:
@libinyang ```type``` should be unused in ipc4 and ```blob_type``` and ```param_id``` unused in ipc3.
@ujfalusi is the kernel using this new ABI IPC4 struct today ?

Reply:
@libinyang can you align this structure in the kernel before we send it upstream, this needs to be the same structure as IPC3 albeit with different magic and use of reserved fields for IPC4 usage. @plbossart fyi.

Reply:
> @ujfalusi is the kernel using this new ABI IPC4 struct today ?

Yes, it is merged via this PR: https://github.com/thesofproject/linux/pull/3918
Unfortunately lots of comments and discussion gone now, especially the ones where I was suggesting to extend/change  the existing sof_abi_hdr struct for IPC4 support.